### PR TITLE
Refactor sandbox preload initialization.

### LINKED ISSUE #12316: Error: Cannot get property 'id' on missing remote object
* Electron version: 2.0.0-beta.4
* Operating system: macOS 10.13.3

### Expected behavior

A reload of the window works.

### Actual behavior

A reload of the window does not work.

### How to reproduce

* run VSCode exploration ([download](https://az764295.vo.msecnd.net/exploration/6bc3274250cc949c7fb1d654cbb6e59c8ddad697/VSCode-darwin-exploration.zip))
* run "Reload Window" command
* sometimes I see the window stay blank after doing so and checking the console I see:

```
Uncaught (in promise) Error: Cannot get property 'id' on missing remote object 3
Error: Cannot get property 'id' on missing remote object 3
    at throwRPCError (/Applications/Visual Studio Code - Exploration.app/Contents/Resources/electron.asar/browser/rpc-server.js:143:17)
    at EventEmitter.<anonymous> (/Applications/Visual Studio Code - Exploration.app/Contents/Resources/electron.asar/browser/rpc-server.js:384:7)
    at emitThree (events.js:136:13)
    at EventEmitter.emit (events.js:217:7)
    at WebContents.<anonymous> (/Applications/Visual Studio Code - Exploration.app/Contents/Resources/electron.asar/browser/api/web-contents.js:295:13)
    at emitTwo (events.js:126:13)
    at WebContents.emit (events.js:214:7)
    at throwRPCError (/Applications/Visual Studio Code - Exploration.app/Contents/Resources/electron.asar/browser/rpc-server.js:143:17)
    at EventEmitter.<anonymous> (/Applications/Visual Studio Code - Exploration.app/Contents/Resources/electron.asar/browser/rpc-server.js:384:7)
    at emitThree (events.js:136:13)
    at EventEmitter.emit (events.js:217:7)
    at WebContents.<anonymous> (/Applications/Visual Studio Code - Exploration.app/Contents/Resources/electron.asar/browser/api/web-contents.js:295:13)
    at emitTwo (events.js:126:13)
    at WebContents.emit (events.js:214:7)
    at Object.exception (/Applications/Visual Studio Code - Exploration.app/Contents/Resources/electron.asar/renderer/api/remote.js:209:30)
    at metaToValue (/Applications/Visual Studio Code - Exploration.app/Contents/Resources/electron.asar/renderer/api/remote.js:213:28)
    at Object.descriptor.get [as id] (/Applications/Visual Studio Code - Exploration.app/Contents/Resources/electron.asar/renderer/api/remote.js:132:16)
    at file:///Applications/Visual%20Studio%20Code%20-%20Exploration.app/Contents/Resources/app/out/vs/workbench/electron-browser/bootstrap/index.js:11:97
    at new Promise (<anonymous>)
    at file:///Applications/Visual%20Studio%20Code%20-%20Exploration.app/Contents/Resources/app/out/vs/workbench/electron-browser/bootstrap/index.js:10:838
```

### PR DESCRIPTION
**Pull request by tarruda**
Use a single synchronous IPC call to retrieve data required by early
sandbox scripts. This has two purposes:

- Optimize preload script initialization by:
  - Using one synchronous IPC call to retrieve preload script,
  webContentsId (more on that later), process.{platform,execPath,env}
  - Lazy loading as many modules as possible.
- Fix #12316 for sandbox. @MarshallOfSound addressed the issue in
  #12342, but it was still present in sandbox mode. By loading
  webContentsId very early and skipping remote module at early
  startup, we fix it for sandbox.

### CODE REVIEW

**Comment by miniak** [comment]
This PR is in conflict with mine https://github.com/electron/electron/pull/12832.

File: lib/browser/rpc-server.js [hunk 1]
```diff
@@ -449,3 +450,13 @@ ipcMain.on('ELECTRON_BROWSER_WINDOW_CLOSE', function (event) {
   }
   event.returnValue = null
 })
+
+ipcMain.on('ELECTRON_BROWSER_SANDBOX_LOAD', function (event, preloadPath) {
```
Comment (miniak) [inline lib/browser/rpc-server.js]: I would not pass the `preloadPath` as an argument. We know which webContents sent the IPC so it should be possible to get the preload script path from the webPreferences.

File: lib/sandboxed_renderer/init.js [hunk 2]
```diff
@@ -43,15 +47,16 @@ preloadProcess.hang = () => binding.hang()
...
+preloadProcess.execPath = process.execPath = execPath
+preloadProcess.env = process.env = env
+
 process.on('exit', () => preloadProcess.emit('exit'))
 
 // This is the `require` function that will be visible to the preload script
 function preloadRequire (module) {
   if (preloadModules.has(module)) {
-    return preloadModules.get(module)
+    return require(module)
```
Comment (miniak) [inline lib/sandboxed_renderer/init.js]: Does this work with browserify?

File: lib/browser/rpc-server.js (hunk 1 above)
Comment (miniak) [reply lib/browser/rpc-server.js]: Apart from this issue I like this change, batching multiple synchronous IPC calls into make to make loading faster is definitely welcome 👍

Comment (miniak) [reply lib/browser/rpc-server.js]: We are actually planning to add some filtering for `remote.require`, `remote.getGlobal`, etc.

Comment (miniak) [reply lib/browser/rpc-server.js]: In my case I had to keep the preloadPath as renderer argument as I need to display an error in case the preload script cannot be loaded. How do you handle that case? Currently we get an error in JS console with the path to the preload script.

Comment (tarruda) [reply lib/browser/rpc-server.js]: > We are actually planning to add some filtering for remote.require, remote.getGlobal, etc.

Cool, that would be a welcome feature :+1: 

> In my case I had to keep the preloadPath as renderer argument as I need to display an error in case the preload script cannot be loaded. How do you handle that case? Currently we get an error in JS console with the path to the preload script.

`ELECTRON_BROWSER_SANDBOX_LOAD` can return an error if it fails to load the script

**Comment by miniak** [comment]
@tarruda off-topic question, why is `clearImmediate` not exposed to the preload script, only `setImmediate`?

File: lib/sandboxed_renderer/init.js [hunk 3]
```diff
@@ -20,19 +20,23 @@ for (let prop of Object.keys(events.EventEmitter.prototype)) {
...
-  ['events', events],
-  ['fs', fs],
-  ['os', require('os')],
-  ['path', require('path')],
-  ['url', require('url')],
-  ['timers', require('timers')]
+const preloadModules = new Set([
+  'child_process',
+  'electron',
+  'fs',
```
Comment (miniak) [inline lib/sandboxed_renderer/init.js]: why did you remove events?

Comment (miniak) [reply lib/sandboxed_renderer/init.js]: It was added recently in https://github.com/electron/electron/pull/12828

File: lib/sandboxed_renderer/init.js (hunk 2 above)
Comment (miniak) [reply lib/sandboxed_renderer/init.js]: What is confusing to me is that we have `fs.js`, `os.js`, `path.js` in `lib/sandboxed_renderer/api/exports`, which are also listed in electron.gyp, but there is no `timers.js` or `url.js`

Comment (miniak) [reply lib/sandboxed_renderer/init.js]: I see, browserify has bundled support for those

Comment (miniak) [reply lib/sandboxed_renderer/init.js]: According to the docs
> Browserify parses the AST for require() calls to traverse the entire dependency graph of your project.

doesn't that mean that all the require calls have to have the module name as literal string?

**Comment by tarruda** [comment]
> @tarruda off-topic question, why is clearImmediate not exposed to the preload script, only setImmediate?

I forgot to add it when this file was created. Will add it now :+1:

File: lib/sandboxed_renderer/init.js (hunk 3 above)
Comment (tarruda) [reply lib/sandboxed_renderer/init.js]: It was a merge error, I took this patch from an older electron version we use in our app. Will add it back :+1:

**Comment by tarruda** [comment]
@miniak about this change:

> In any case, I agree that it would be cleaner to take preloadPath from webPreferences directly as it would also remove the need to pass the path as a renderer argument.

Since it involves some C++ changes I will do it in a separate PR ok?

File: lib/browser/rpc-server.js (hunk 1 above)
Comment (miniak) [reply lib/browser/rpc-server.js]: Returning an error would be great. We should not have a silent failure without any logs in the console.

Comment (miniak) [reply lib/browser/rpc-server.js]: Also if you just let fs.readFileSync throw and not assign the returnValue, sendSync in the renderer process will get stuck waiting.

File: lib/sandboxed_renderer/init.js [hunk 4]
```diff
@@ -43,15 +51,19 @@ preloadProcess.hang = () => binding.hang()
...
 
 // This is the `require` function that will be visible to the preload script
 function preloadRequire (module) {
-  if (preloadModules.has(module)) {
-    return preloadModules.get(module)
+  if (loadedModules.has(module)) {
+    return loadedModules.get(module)
+  }
+  if (remoteModules.has(module)) {
+    return require(module)
```
Comment (miniak) [inline lib/sandboxed_renderer/init.js]: shouldn't this be electron.remote.require instead?

Comment (tarruda) [reply lib/sandboxed_renderer/init.js]: No, we still require these files to emulate a real node.js require for electron renderer modules that are reused in sandbox. Those are bundled in the executable and don't know they are running in a browserify environment.

File: lib/sandboxed_renderer/init.js [hunk 5]
```diff
@@ -20,19 +21,26 @@ for (let prop of Object.keys(events.EventEmitter.prototype)) {
...
-const preloadModules = new Map([
-  ['child_process', require('child_process')],
+const remoteModules = new Set([
+  'child_process',
+  'fs',
+  'os',
+  'path'
+])
+
+const loadedModules = new Map([
```
[the remark this answers is not in the document]
Comment (tarruda) [reply lib/sandboxed_renderer/init.js]: Technically, all modules are bundled. The ones that delegate to electron.remote are loaded lazily, but are still bundled.

File: lib/sandboxed_renderer/init.js (hunk 4 above)
Comment (tarruda) [reply lib/sandboxed_renderer/init.js]: What I mean is, electron modules such as crashReporter won't use this `require` function, so we need to tell browserify that when crashReporter requires `fs`, it is requiring `/lib/sandboxed_renderer/api/exports/fs.js` and not browserify shim.

**Comment by tarruda** [comment]
@miniak With exception of taking preload path from webPreferences (which I want to do in a separate PR), I believe I've addressed all your comments. Let me know when you've finished your review so I can squash the fixup commits.

File: lib/sandboxed_renderer/init.js (hunk 4 above)
Comment (miniak) [reply lib/sandboxed_renderer/init.js]: ok

File: lib/sandboxed_renderer/init.js [hunk 6]
```diff
@@ -80,16 +92,15 @@ if (window.location.protocol === 'chrome-devtools:') {
...
   ${preloadSrc}
   })`
 
   // eval in window scope:
   // http://www.ecma-international.org/ecma-262/5.1/#sec-10.4.2
   const geval = eval
   const preloadFn = geval(preloadWrapperSrc)
-  const {setImmediate} = require('timers')
-  preloadFn(preloadRequire, preloadProcess, Buffer, global, setImmediate)
+  const {setImmediate, clearImmediate} = require('timers')
```
Comment (miniak) [inline lib/sandboxed_renderer/init.js]: 👍

File: lib/sandboxed_renderer/init.js [hunk 7]
```diff
@@ -80,16 +92,15 @@ if (window.location.protocol === 'chrome-devtools:') {
 // and any `require('electron')` calls in `preload.js` will work as expected
 // since browserify won't try to include `electron` in the bundle, falling back
 // to the `preloadRequire` function above.
-if (preloadPath) {
-  const preloadSrc = fs.readFileSync(preloadPath).toString()
-  const preloadWrapperSrc = `(function(require, process, Buffer, global, setImmediate) {
+if (preloadSrc) {
```
Comment (miniak) [inline lib/sandboxed_renderer/init.js]: you need to check here if preloadSrc is an Error

File: lib/browser/rpc-server.js [hunk 8]
```diff
@@ -449,3 +450,21 @@ ipcMain.on('ELECTRON_BROWSER_WINDOW_CLOSE', function (event) {
...
+  let preloadSrc = null
+  if (preloadPath) {
+    try {
+      preloadSrc = fs.readFileSync(preloadPath).toString()
+    } catch (err) {
+      preloadSrc = err
```
Comment (miniak) [inline lib/browser/rpc-server.js]: maybe a separate variable? preloadError?

Comment (MarshallOfSound) [reply lib/browser/rpc-server.js]: Also we need to be aware the `Error` objects to not JSON stringify very well by default, you'll need to manually pull out the `stack` and `message` attributes into a new object

File: lib/browser/rpc-server.js (hunk 1 above)
Comment (MarshallOfSound) [reply lib/browser/rpc-server.js]: I agree with @miniak this is basically a one-liner "read any file plz" API, I'd rather we (in this PR) used `event.sender.getLastWebPreferences()`

File: lib/sandboxed_renderer/init.js [hunk 9]
```diff
@@ -20,19 +21,26 @@ for (let prop of Object.keys(events.EventEmitter.prototype)) {
...
-  ['timers', require('timers')]
+  ['timers', require('timers')],
+  ['url', require('url')]
 ])
 
+const {
+  preloadSrc, webContentsId, platform, execPath, env
+} = electron.ipcRenderer.sendSync('ELECTRON_BROWSER_SANDBOX_LOAD', preloadPath)
+
+process.webContentsId = webContentsId
```
Comment (MarshallOfSound) [inline lib/sandboxed_renderer/init.js]: For safety

```js
Object.defineProperty(process, 'webContentsId', {
  configurable: false,
  value: webContentsId
})
```

This way nothing can ever override that value

File: lib/sandboxed_renderer/init.js [hunk 10]
```diff
@@ -103,4 +106,6 @@ if (preloadSrc) {
   const preloadFn = geval(preloadWrapperSrc)
   const {setImmediate, clearImmediate} = require('timers')
   preloadFn(preloadRequire, preloadProcess, Buffer, global, setImmediate, clearImmediate)
+} else if (preloadError) {
+  console.error(preloadError.stack)
```
Comment (tarruda) [inline lib/sandboxed_renderer/init.js]: @miniak is this how you wanted to handle preload error? I can't think of a better way right now, logging via `console.error` seems like a sane way to do it, since it would be available in the terminal if electron is started with `--enable-logging`

File: lib/browser/rpc-server.js [hunk 11]
```diff
@@ -449,3 +450,23 @@ ipcMain.on('ELECTRON_BROWSER_WINDOW_CLOSE', function (event) {
...
 })
+
+ipcMain.on('ELECTRON_BROWSER_SANDBOX_LOAD', function (event, preloadPath) {
+  let preloadSrc = null
+  let preloadError = null
+  if (preloadPath) {
+    try {
+      preloadSrc = fs.readFileSync(preloadPath).toString()
+    } catch (err) {
+      preloadError = {stack: err.stack}
```
Comment (miniak) [inline lib/browser/rpc-server.js]: why don't you include name and message as well?

File: lib/sandboxed_renderer/init.js (hunk 7 above)
Comment (miniak) [reply lib/sandboxed_renderer/init.js]: you are not logging the error here

Comment (tarruda) [reply lib/sandboxed_renderer/init.js]: I've used  `preloadError` for the error

File: lib/browser/rpc-server.js (hunk 11 above)
Comment (tarruda) [reply lib/browser/rpc-server.js]: The `stack` property already contains the message and Error class name. Example from the repl:

```
> console.log(new Error('msg').stack)
Error: msg
    at repl:1:13
    at ContextifyScript.Script.runInThisContext (vm.js:50:33)
    at REPLServer.defaultEval (repl.js:240:29)
    at bound (domain.js:301:14)
    at REPLServer.runBound [as eval] (domain.js:314:12)
    at REPLServer.onLine (repl.js:468:10)
    at emitOne (events.js:121:20)
    at REPLServer.emit (events.js:211:7)
    at REPLServer.Interface._onLine (readline.js:282:10)
    at REPLServer.Interface._line (readline.js:631:8)
```

File: lib/browser/rpc-server.js (hunk 1 above)
Comment (miniak) [reply lib/browser/rpc-server.js]: @tarruda this is the last remaining issue

File: lib/browser/rpc-server.js (hunk 11 above)
Comment (MarshallOfSound) [inline lib/browser/rpc-server.js]: I hate to be that person 😆 But welcome to javascript where you can `throw null`

This should check if we can get to `.stack` before actually trying or our error handling will error 😆

File: lib/browser/rpc-server.js (hunk 1 above)
Comment (tarruda) [reply lib/browser/rpc-server.js]: > I agree with @miniak this is basically a one-liner "read any file plz" API, I'd rather we (in this PR) used event.sender.getLastWebPreferences()

We currently give access to `electron.remote.fs` which can not only read any file, but also invoke any FS function. There's no regression in terms of security here by allowing the renderer to specify the preloadPath.

These are my arguments for keeping the current behavior and addressing this issue in a separate PR:

- If we start using `event.sender.getLastWebPreferences()`, there's no  reason to pass the preload path as argument to the renderer command line, so it would become necessary to refactor more code in C++ both in browser/renderer. (Unless we are OK with leaving dead code to be handled in a separate PR)
- There's no guarantee that this will work as expected without introducing new regressions. Maybe it will, but I will have to investigate if `preload` will be present in every `event.sender.webPreferences` (in windows created by window.open, for example, which follow a different webContents creation path).
- This is a completely separate change, not related to the goal of this PR which is to fix  #12316. I'd rather push refactoring changes in different PRs.

If you guys still insist that this should be done in the current PR, I will do it, but not right now since it will require more testing/verification which I can't do this week.

File: lib/browser/rpc-server.js (hunk 11 above)
Comment (tarruda) [reply lib/browser/rpc-server.js]: :+1:

**Comment by tarruda** [comment]
Rebased/squashed commits

File: lib/browser/rpc-server.js (hunk 1 above)
Comment (miniak) [reply lib/browser/rpc-server.js]: @MarshallOfSound? I guess I am ok if this gets addressed in a separate PR.

File: lib/sandboxed_renderer/init.js [hunk 12]
```diff
@@ -20,19 +21,29 @@ for (let prop of Object.keys(events.EventEmitter.prototype)) {
...
+  ['timers', require('timers')],
+  ['url', require('url')]
 ])
 
+const {
+  preloadSrc, preloadError, webContentsId, platform, execPath, env
+} = electron.ipcRenderer.sendSync('ELECTRON_BROWSER_SANDBOX_LOAD', preloadPath)
+
+Object.defineProperty(process, 'webContentsId', {
+  configurable: false,
```
Comment (alexeykuzmin) [inline lib/sandboxed_renderer/init.js]: `false` is a default value of the `configurable` attribute, no reasons to explicitly set it.
If you want to be explicit then you should also set `writable: false` (which is also a default value).

**Comment by alexeykuzmin** [comment]
@tarruda Can you please backport the changes to `2-0-x`?

**Comment by MarshallOfSound** [comment]
cc @tarruda ^^

**Comment by tarruda** [comment]
:+1:
